python/tk_multi_publish2/api/data.py

```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import collections
+
+
+class PublishData(collections.MutableMapping):
+    """
+    A simple dictionary-like object for storing arbitrary data.
+
+    Provides access via standard dict syntax as well as dot notation. This is
+    used as the base class for any arbitrary data exposed by the publish API
+    including internal representation of settings (as configured or modified
+    by the UI) and publish item properties.
+    """
+
+    @classmethod
+    def clone(cls, pub_data_obj):
```

Comment:
I've always preferred classmethods for this type of thing, but maybe an instance method is more idiomatic. I'll do some searching to see what's more preferable.

Reply:
Maybe it's because I'm coming at this from a C++ background.

Having a method implemented in each class called clone() that returns a pointer to the cloned object is pretty standard fare. The only difference here is that in Python we can get away with writing a single method, which is pretty nice.

Reply:
besides

```
item = item.clone() 
```

feels nicer to me than

```
item = PublishData.clone(item)
```

Reply:
Come to thing of if, shouldn't we have implemented this as
return copy.deepcopy(self)?